File: client/appearance/apparatusDatabase.lua

function getHumanDatabase()
    if gMpPedsComponents then
        return gMpPedsComponents
    end

    return {}
end
    

function getMetapedBodyApparatusFromShopitemAny(apparatusType, apparatusGender, apparatusId, apparatusStyleId, shopitemHash)
    local dbLayerRoot = getHumanDatabase()

    local function found(type, gender, id, styleId)
        return {
            id = tonumber(id),
            styleId = tonumber(styleId),
            type = tonumber(type),
            gender = tonumber(gender)
        }
    end

    if shopitemHash == nil then
        return
    end

    if apparatusType ~= nil then
        local dbLayerType = dbLayerRoot[apparatusType]

        if apparatusGender ~= nil then
            local dbLayerGender = dbLayerType[apparatusGender]

            if apparatusId ~= nil then
                local dbLayerStyle = dbLayerGender[apparatusId]

                if apparatusStyleId ~= nil then
                    local shopitemAny = dbLayerStyle[apparatusStyleId]
                    return found(apparatusType, apparatusGender, apparatusId, apparatusStyleId)
                else
                    for itApparatusStyleId, itShopitemAny in pairs(dbLayerStyle) do
                        if itShopitemAny == shopitemHash or GetHashKey(itShopitemAny) == shopitemHash then
                            return found(apparatusType, apparatusGender, apparatusId, itApparatusStyleId)
                        end
                    end
                end
            else
                for itApparatusId, itDbLayerStyle in pairs(dbLayerGender) do
                    if apparatusStyleId ~= nil then
                        if itDbLayerStyle[apparatusStyleId] == shopitemHash then
                            return found(apparatusType, apparatusGender, itApparatusId, apparatusStyleId)
                        end
                    else
                        for itApparatusStyleId, itShopitemAny in pairs(itDbLayerStyle) do
                            if itShopitemAny == shopitemHash or GetHashKey(itShopitemAny) == shopitemHash then
                                return found(apparatusType, apparatusGender, itApparatusId, itApparatusStyleId)
                            end
                        end
                    end
                end
            end
        else
            for itApparatusGender, itDbLayerGender in pairs(dbLayerType) do
                if apparatusId ~= nil then
                    local dbLayerStyle = itDbLayerGender[apparatusId]

                    if apparatusStyleId ~= nil then
                        local shopitemAny = dbLayerStyle[apparatusStyleId]

                        if shopitemAny == shopitemHash then
                            return found(apparatusType, itApparatusGender, apparatusId, apparatusStyleId)
                        end
                    else
                        for itApparatusStyleId, itShopitemAny in pairs(dbLayerStyle) do
                            if itShopitemAny == shopitemHash or GetHashKey(itShopitemAny) == shopitemHash then
                                return found(apparatusType, itApparatusGender, apparatusId, itApparatusStyleId)
                            end
                        end
                    end
                else
                    for itApparatusId, itDbLayerStyle in pairs(itDbLayerGender) do
                        if apparatusStyleId ~= nil then
                            if itDbLayerStyle[apparatusStyleId] == shopitemHash then
                                return found(apparatusType, itApparatusGender, itApparatusId, apparatusStyleId)
                            end
                        else
                            for itApparatusStyleId, itShopitemAny in pairs(itDbLayerStyle) do
                                if itShopitemAny == shopitemHash or GetHashKey(itShopitemAny) == shopitemHash then
                                    return found(apparatusType, itApparatusGender, itApparatusId, itApparatusStyleId)
                                end
                            end
                        end
                    end
                end
            end
        end
    else
        for itApparatusType, itDbLayerType in pairs(dbLayerRoot) do
            if apparatusGender ~= nil then
                local dbLayerGender = itDbLayerType[apparatusGender]

                if apparatusId ~= nil then
                    local dbLayerStyle = dbLayerGender[apparatusId]

                    if apparatusStyleId ~= nil then
                        return found(itApparatusType, apparatusGender, apparatusId, apparatusStyleId)
                    else
                        for itApparatusStyleId, itShopitemAny in pairs(dbLayerStyle) do
                            if itShopitemAny == shopitemHash or GetHashKey(itShopitemAny) == shopitemHash then
                                return found(itApparatusType, apparatusGender, apparatusId, itApparatusStyleId)
                            end
                        end
                    end
                else
                    for itApparatusId, itDbLayerStyle in pairs(dbLayerGender) do
                        if apparatusStyleId ~= nil then
                            if itDbLayerStyle[apparatusStyleId] == shopitemHash then
                                return found(itApparatusType, apparatusGender, itApparatusId, apparatusStyleId)
                            end
                        else
                            for itApparatusStyleId, itShopitemAny in pairs(itDbLayerStyle) do
                                if itShopitemAny == shopitemHash or GetHashKey(itShopitemAny) == shopitemHash then
                                    return found(itApparatusType, apparatusGender, itApparatusId, itApparatusStyleId)
                                end
                            end
                        end
                    end
                end
            else
                for itApparatusGender, itDbLayerGender in pairs(itDbLayerType) do
                    if apparatusId ~= nil then
                        local dbLayerStyle = itDbLayerGender[apparatusId]

                        if apparatusStyleId ~= nil then
                            return found(itApparatusType, itApparatusGender, apparatusId, apparatusStyleId)
                        end
                    else
                        for itApparatusId, itDbLayerStyle in pairs(itDbLayerGender) do
                            if apparatusStyleId ~= nil then
                                local itShopitemAny = itDbLayerStyle[itApparatusId]

                                if itShopitemAny == shopitemHash or GetHashKey(itShopitemAny) == shopitemHash then
                                    return found(itApparatusType, itApparatusGender, itApparatusId, apparatusStyleId)
                                end
                            else
                                for itApparatusStyleId, itShopitemAny in pairs(itDbLayerStyle) do
                                    if itShopitemAny == shopitemHash or GetHashKey(itShopitemAny) == shopitemHash then
                                        return found(itApparatusType, itApparatusGender, itApparatusId, itApparatusStyleId)
                                    end
                                end
                            end
                        end
                    end
                end
            end
        end
    end
end

function getShopitemAnyByMetapedBodyApparatus(metapedBodyApparatus)
    local dbLayerRoot = getHumanDatabase()
    -- print(" dbLayerRoot :: ", dbLayerRoot)

    local type = metapedBodyApparatus.type
    local gender = metapedBodyApparatus.gender
    local id = metapedBodyApparatus.id
    local styleId = metapedBodyApparatus.styleId

    -- print("  metapedBodyApparatus ", type, gender, id, styleId)
    local success, errorMessage

    local dbLayerType = dbLayerRoot[type + 1]

    if dbLayerType ~= nil then
        local dbLayerGender = dbLayerType[gender]

        if dbLayerGender ~= nil then
            local dbLayerStyle = dbLayerGender[id]

            if dbLayerStyle ~= nil then
                local shopitem = dbLayerStyle[styleId]

                if shopitem ~= nil then
                    return shopitem
                end
            end
        end
    end

end


function isMetapedUsingApparatusRDR3(pedId, apparatus)
    local shopitem = getShopitemAnyByMetapedBodyApparatus(apparatus);

    if not shopitem then
        return false;
    end

    local shopitemHash = type(shopitem) == 'string' and GetHashKey(shopitem) or shopitem;

    -- Não funciona por algum motivo :/ talvez só funcione com componentGroups.
    -- IsMetapedUsingComponent
    -- return !!N_0xfb4891bd7578cdc1(pedId, shopitemHash);

    -- GetNumComponentsInPed | _GetNumMetapedAssets
    local numMetaAssets = N_0x90403e8107b60e81(pedId); 

    if numMetaAssets <= 0 then
        return false;
    end

    local buffer = DataView.ArrayBuffer(8 * 5);
    local view = buffer:Buffer()

    for i = 0, numMetaAssets - 1 do
        -- GetPedComponentAtIndex
        local itShopitemHash = N_0x77ba37622e22023b(pedId, i, true, view, view);

        if shopitemHash == itShopitemHash then
            return true;
        end
    end

    return false;
end

File: client/appearance-js/apparatusDatabase.js
// RegisterCommand("startPed", (_, args) => {
//     const playerPed = PlayerPedId();
//     handleStartEditor(playerPed);
// }, false);
// RegisterCommand("cc", (_, args) => {
//     const id1 = args[0];
//     const id2 = args[1];
//     const id3 = args[2];

//     const playerPed = PlayerPedId();

//     requestChangeApparatus(playerPed, {
//         component: id1,
//         data: Number(id2),
//         type: id3
//     });
// }, false);

async function getHumanDatabase()
{
    const dataFile = global.exports.frp_core.mp_peds_components();
    if (dataFile)
    {
        return dataFile
    }
}

// console.log(`equippedApparels`, equippedApparels);
async function getShopitemAnyByMetapedBodyApparatus(metapedBodyApparatus)
{
    const dbLayerRoot = await getHumanDatabase();
    // console.log(" dbLayerRoot :: ", dbLayerRoot);

    const { type, gender, id, styleId } = metapedBodyApparatus;

    // console.log("  metapedBodyApparatus ", type, gender, id, styleId)
    try
    {
        const dbLayerType   = dbLayerRoot[type];

        // if (!dbLayerType) throw new ApplicationException('Type');

        const dbLayerGender = dbLayerType[gender];

        // if (!dbLayerGender) throw new ApplicationException('Gender');

        const dbLayerStyle  = dbLayerGender[id];

        // if (!dbLayerStyle) throw new ApplicationException('Style');

        const shopitem = dbLayerStyle[styleId];

        return shopitem;
    }
    catch(e)
    {
        // throw new ApplicationException(`Não foi possível entrar um Shopitem para o Apparatus: Tipo(${eMetapedBodyApparatusType[type] ?? type}) Genêro(${eMetapedBodyApparatusGender[gender] ?? gender}) Id(${id}) Style(${styleId}) Layer(${e.message})`);
    }
};

async function getMetapedBodyApparatusFromShopitemAny(apparatusType, apparatusGender, apparatusId, apparatusStyleId, shopitemHash)
{
    const dbLayerRoot = await getHumanDatabase();

    const found = (type, gender, id, styleId) =>
    {
        return {
            id: Number(id),
            styleId: Number(styleId),
            type: Number(type),
            gender: Number(gender),
        };
    };

    if (shopitemHash == null)
    {
        return;
    }
    
    if (apparatusType != null)
    {
        const dbLayerType = dbLayerRoot[apparatusType];

        if (apparatusGender != null)
        {
            const dbLayerGender = dbLayerType[apparatusGender]

            if (apparatusId != null)
            {
                const dbLayerStyle = dbLayerGender[apparatusId];

                // console.log('apparatusStyleId', apparatusStyleId);

                if (apparatusStyleId != null)
                {
                    const shopitemAny = dbLayerStyle[apparatusStyleId];

                    // return [ 'yes', -1 ];

                    return found(apparatusType, apparatusGender, apparatusId, apparatusStyleId);
                }
                else
                {
                    for (const [itApparatusStyleId, itShopitemAny] of Object.entries(dbLayerStyle))
                    {
                        if (itShopitemAny == shopitemHash || GetHashKey(itShopitemAny) == shopitemHash)
                        {
                            // return [ 'yes', 0 ];

                            return found(apparatusType, apparatusGender, apparatusId, itApparatusStyleId);
                        }
                    }
                }
            }
            else
            {
                for (const [itApparatusId, itDbLayerStyle] of Object.entries(dbLayerGender))
                {
                    if (apparatusStyleId != null)
                    {
                        if (itDbLayerStyle[apparatusStyleId] == shopitemHash)
                        {
                            // return [ 'yes', 1 ];

                            return found(apparatusType, apparatusGender, itApparatusId, apparatusStyleId);
                        }
                    }
                    else
                    {
                        for (const [itApparatusStyleId, itShopitemAny] of Object.entries(itDbLayerStyle))
                        {
                            if (itShopitemAny == shopitemHash || GetHashKey(itShopitemAny ) == shopitemHash)
                            {
                                // return [ 'yes', 2 ];

                                return found(apparatusType, apparatusGender, itApparatusId, itApparatusStyleId);
                            }
                        } 
                    }
                }
            }
        }
        else
        {
            for (const [ítApparatusGender, itDbLayerGender] of Object.entries(dbLayerType))
            {
                if (apparatusId != null)
                {
                    const dbLayerStyle = itDbLayerGender[apparatusId];

                    if (apparatusStyleId != null)
                    {
                        const shopitemAny = dbLayerStyle[apparatusStyleId]

                        if (shopitemAny == shopitemHash)
                        {
                            // return [ 'yes', 3 ];

                            return found(apparatusType, ítApparatusGender, apparatusId, apparatusStyleId);
                        }
                    }
                    else
                    {
                        for (const [itApparatusStyleId, itShopitemAny] of Object.entries(dbLayerStyle))
                        {
                            if (itShopitemAny == shopitemHash || GetHashKey(itShopitemAny ) == shopitemHash)
                            {
                                // return [ 'yes', 4 ];

                                return found(apparatusType, ítApparatusGender, apparatusId, itApparatusStyleId);
                            }
                        }
                    }
                }
                else
                {
                    for (const [itApparatusId, itDbLayerStyle] of Object.entries(itDbLayerGender))
                    {
                        if (apparatusStyleId != null)
                        {
                            const shopitemAny = itDbLayerStyle[apparatusStyleId];

                            if (shopitemAny == shopitemHash)
                            {
                                // return [ 'yes', 5 ];

                                return found(apparatusType, ítApparatusGender, itApparatusId, apparatusStyleId);
                            }
                        }
                        else
                        {
                            for (const [itApparatusStyleId, itShopitemAny] of Object.entries(itDbLayerStyle))
                            {
                                if (itShopitemAny == shopitemHash || GetHashKey(itShopitemAny) == shopitemHash)
                                {
                                    // return [ 'yes', 6 ];

                                    return found(apparatusType, ítApparatusGender, itApparatusId, itApparatusStyleId);
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    else
    {
        for (const [itApparatusType, itDbLayerType] of Object.entries(dbLayerRoot))
        {
            if (apparatusGender != null)
            {
                const dbLayerGender = itDbLayerType[apparatusGender];

                if (apparatusId != null)
                {
                    const dbLayerStyle = dbLayerGender[apparatusId];

                    if (apparatusStyleId != null)
                    {
                        // shopitemAny = dbLayerStyle[apparatusStyleId]

                        // return [ 'yes', 7 ];

                        return found(itApparatusType, apparatusGender, apparatusId, apparatusStyleId);
                    }
                    else
                    {
                        for (const [itApparatusStyleId, itShopitemAny] of Object.entries(dbLayerStyle))
                        {
                            if (itShopitemAny == shopitemHash || GetHashKey(itShopitemAny) == shopitemHash)
                            {
                                // return [ 'yes', 8 ];

                                return found(itApparatusType, apparatusGender, apparatusId, itApparatusStyleId);
                            }
                        }
                    }
                }
                else
                {
                    for (const [itApparatusId, itDbLayerStyle] of Object.entries(dbLayerGender))
                    {
                        if (apparatusStyleId != null)
                        {
                            if (itDbLayerStyle[apparatusStyleId] == shopitemHash)
                            {
                                // return [ 'yes', 9 ];

                                return found(itApparatusType, apparatusGender, itApparatusId, apparatusStyleId);
                            }
                        }
                        else
                        {
                            for (const [itApparatusStyleId, itShopitemAny] of Object.entries(itDbLayerStyle))
                            {
                                if (itShopitemAny == shopitemHash || GetHashKey(itShopitemAny) == shopitemHash)
                                {
                                    // return [ 'yes', 10 ];

                                    return found(itApparatusType, apparatusGender, itApparatusId, itApparatusStyleId);
                                }
                            }
                        }
                    }
                }
            }
            else
            {
                for (const [itApparatusGender, itDbLayerGender] of Object.entries(itDbLayerType))
                {
                    if (apparatusId != null)
                    {
                        const dbLayerStyle = itDbLayerGender[apparatusId];

                        if (apparatusStyleId != null)
                        {
                            // shopitemAny = dbLayerStyle[apparatusStyleId];

                            // return [ 'yes', 11 ];

                            return found(itApparatusType, itApparatusGender, apparatusId, apparatusStyleId);
                        }
                    }
                    else
                    {
                        for (const [itApparatusId, itDbLayerStyle] of Object.entries(itDbLayerGender))
                        {
                            if (apparatusStyleId != null)
                            {
                                const itShopitemAny = itDbLayerStyle[itApparatusId];

                                if (itShopitemAny == shopitemHash || GetHashKey(itShopitemAny ) == shopitemHash)
                                {
                                    // return [ 'yes', 12 ];

                                    return found(itApparatusType, itApparatusGender, itApparatusId, apparatusStyleId);
                                }
                            }
                            else
                            {
                                for (const [itApparatusStyleId, itShopitemAny] of Object.entries(itDbLayerStyle))
                                {
                                    if (itShopitemAny == shopitemHash || GetHashKey(itShopitemAny ) == shopitemHash)
                                    {
                                        // return [ 'yes', 13 ];

                                        return found(itApparatusType, itApparatusGender, itApparatusId, itApparatusStyleId);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}
exports('getMetapedBodyApparatusFromShopitemAny', getMetapedBodyApparatusFromShopitemAny)

async function isMetapedUsingApparatusRDR3(pedId, apparatus)
{
    const shopitem = await getShopitemAnyByMetapedBodyApparatus(apparatus);

    if (!shopitem)
    {
        return false;
    }

    const shopitemHash = typeof shopitem == 'string' ? GetHashKey(shopitem) : shopitem;

    // Não funciona por algum motivo :/ talvez só funcione com componentGroups.
    // IsMetapedUsingComponent
    // return !!N_0xfb4891bd7578cdc1(pedId, shopitemHash);

    // GetNumComponentsInPed | _GetNumMetapedAssets
    const numMetaAssets = N_0x90403e8107b60e81(pedId); 

    if (numMetaAssets <= 0)
    {
        return false;
    }

    const buffer = new ArrayBuffer(8 * 5);
    const view = new DataView(buffer);

    for (let i = 0; i < numMetaAssets; i++)
    {
        // GetPedComponentAtIndex
        const itShopitemHash = N_0x77ba37622e22023b(pedId, i, true, view, view);

        if (shopitemHash == itShopitemHash)
        {
            return true;
        }
    }

    return false;
};